File: python/model_if.py
# Test ov::Node::If

from openvino.runtime import Core, Model, Tensor, PartialShape, Type, Shape, op, serialize
import openvino.runtime as ov
import openvino.properties.hint as hints
from openvino.runtime.op import util as op_util
from openvino.runtime import opset8 as opset
from openvino.runtime.passes import Manager
import numpy as np
import time
import os

# ======== Model graph ========
#              input1
#                 |
# intput2        If     input3
#    |          /   \      |
#  then(multiply1)  else(multiply2)
#               \   /
#                Add
#                 |
#               Output
def model_if():
    input1 = opset.parameter([1], Type.i32, "input1")
    input_for_then = opset.parameter([1], Type.i32, "input_for_then")
    input_for_else = opset.parameter([1], Type.i32, "input_for_else")

    # equal = ov.opset8.equal(input1.output(0), op.Constant(np.full((1), 2).astype(np.int32)))
    if_op = ov.opset8.if_op()

    input_then = ov.opset8.parameter([1], np.int32, "input_then")
    multiply1 = opset.multiply(input_then, op.Constant(np.full((1), 2).astype(np.int32)))
    multiply1.set_friendly_name("multiply1")
    then_op_result = ov.opset6.result(multiply1, "res_then")
    body_then_function = Model(results=[then_op_result], parameters=[input_then], name='model_then_body')

    input_else = ov.opset8.parameter([1], np.int32, "input_else")
    multiply2 = opset.multiply(input_else, op.Constant(np.full((1), 4).astype(np.int32)))
    multiply2.set_friendly_name("multiply2")
    else_op_result = ov.opset6.result(multiply2, "res_else")
    body_else_function = Model(results=[else_op_result], parameters=[input_else], name='body_else_function')

    if_op.set_then_body(body_then_function)
    if_op.set_else_body(body_else_function)

    if_op.set_input(input1.output(0), input_then, input_else)
    if_op.set_input(input_for_then.output(0), input_then, None)
    if_op.set_input(input_for_else.output(0), None, input_else)

    result_if = if_op.set_output(then_op_result, else_op_result)

    add = opset.add(result_if, op.Constant(np.full((1), 0).astype(np.int32)))
    res = ov.opset6.result(add, "res")
    return Model(results=[res], parameters=[input1, input_for_then, input_for_else], name='model_if')

def test_model_if(device:str, then_branch=True):
    print(f'== test_model_if device = {device}, then_branch = {then_branch}')
    core = Core()
    model = model_if()
    compiled_model = core.compile_model(model=model, device_name=device)
    
    input1=np.array([1] if then_branch else [0]).astype(np.float32)

    input_for_then=np.array([2]).astype(np.int32)
    input_for_else=np.array([4]).astype(np.int32)
    infer_request = compiled_model.create_infer_request()

    print(f"== Run model_if, device={device}")
    result = infer_request.infer([input1, input_for_then, input_for_else])[compiled_model.output(0)]
    print(f'== reuslt = {result}, expected = {input_for_then*2 if then_branch else input_for_else*4}')


# ======== Model graph ========
#              input1
#                 |
# input2/input3  If   input4/input5
#    |          /   \      |
#  then(multiply1)  else(multiply2)
#               \   /
#                Add
#                 |
#               Output
def model_if_multiple_inputs():
    input1 = opset.parameter([1], Type.i32, "input1")
    input_for_then_1 = opset.parameter([1], Type.i32, "input_for_then_1")
    input_for_then_2 = opset.parameter([1], Type.i32, "input_for_then_2")
    input_for_else_1 = opset.parameter([1], Type.i32, "input_for_else_1")
    input_for_else_2 = opset.parameter([1], Type.i32, "input_for_else_2")

    if_op = ov.opset8.if_op()
    # then:
    input_then_1 = ov.opset8.parameter([1], np.int32, "input_then_1")
    input_then_2 = ov.opset8.parameter([1], np.int32, "input_then_2")
    multiply1 = opset.multiply(input_then_1, op.Constant(np.full((1), 2).astype(np.int32)))
    multiply1.set_friendly_name("multiply1")
    add = opset.add(multiply1, input_then_2)
    then_op_result = ov.opset6.result(add, "res_then")
    body_then_function = Model(results=[then_op_result], parameters=[input_then_1, input_then_2], name='model_then_body')

    # else:
    input_else_1 = ov.opset8.parameter([1], np.int32, "input_else_1")
    input_else_2 = ov.opset8.parameter([1], np.int32, "input_else_2")
    multiply_1 = opset.multiply(input_else_1, op.Constant(np.full((1), 4).astype(np.int32)))
    multiply_1.set_friendly_name("multiply2")
    add_1 = opset.add(multiply_1, input_else_2)
    else_op_result = ov.opset6.result(add_1, "res_else")
    body_else_function = Model(results=[else_op_result], parameters=[input_else_1, input_else_2], name='body_else_function')

    if_op.set_then_body(body_then_function)
    if_op.set_else_body(body_else_function)

    if_op.set_input(input1.output(0), input_then_1, input_else_1)
    if_op.set_input(input_for_then_1.output(0), input_then_1, None)
    if_op.set_input(input_for_then_2.output(0), input_then_2, None)
    if_op.set_input(input_for_else_1.output(0), None, input_else_1)
    if_op.set_input(input_for_else_2.output(0), None, input_else_2)

    result_if = if_op.set_output(then_op_result, else_op_result)

    add = opset.add(result_if, op.Constant(np.full((1), 0).astype(np.int32)))
    res = ov.opset6.result(add, "res")
    return Model(results=[res], parameters=[input1, input_for_then_1, input_for_then_2, input_for_else_1, input_for_else_2], name='model_if')

def test_model_if_multiple_inputs(device:str, then_branch=True):
    print(f'== test_model_if_multiple_inputs device = {device}, then_branch = {then_branch}')
    core = Core()
    model = model_if_multiple_inputs()
    compiled_model = core.compile_model(model=model, device_name=device)
    
    input1=np.array([1] if then_branch else [0]).astype(np.float32)

    input_for_then_1=np.array([2]).astype(np.int32)
    input_for_then_2=np.array([3]).astype(np.int32)
    input_for_else_1=np.array([4]).astype(np.int32)
    input_for_else_2=np.array([3]).astype(np.int32)
    infer_request = compiled_model.create_infer_request()

    print(f"== Run model_if, device={device}")
    result = infer_request.infer([input1, input_for_then_1, input_for_then_2, input_for_else_1, input_for_else_2])[compiled_model.output(0)]
    print(f'== reuslt = {result}, expected = {input_for_then_1*2+3 if then_branch else input_for_else_1*4+3}')

# == Test ReadValue's input is parameter
if __name__ == "__main__":
    print(f'ov version:{ov.get_version()}')
    # test_model_if('TEMPLATE')
    # test_model_if('CPU')
    # test_model_if('GPU', then_branch=True)
    # test_model_if('GPU', then_branch=False)
    test_model_if_multiple_inputs('GPU', then_branch=True)
    test_model_if_multiple_inputs('GPU', then_branch=False)
    

File: cpp/src/model_if.cpp
#include "private.hpp"
#include "utils.hpp"

#include <openvino/openvino.hpp>
#include <openvino/opsets/opset1.hpp>
#include <openvino/opsets/opset6.hpp>
#include <openvino/opsets/opset8.hpp>

// Construct OpenVINO subgraph
// Input1      const
//   \           |
//    \       convert
//     \         |
//      \    ReadValue
//       \      /
//        MatMul
//          |
//        Output
static std::shared_ptr<ov::Model> initIfModel()
{
    std::cout << " == Init model: ReadValue inpupt with const" << std::endl;
    // Input 1: parameter
    auto input1 = std::make_shared<ov::opset1::Parameter>(ov::element::i32, ov::Shape({1}));
    auto input2 = std::make_shared<ov::opset1::Parameter>(ov::element::i32, ov::Shape({1}));
    auto input3 = std::make_shared<ov::opset1::Parameter>(ov::element::i32, ov::Shape({1}));

    // If
    auto if_op = std::make_shared<ov::opset8::If>();

    // If then
    auto inp_then = std::make_shared<ov::opset1::Parameter>(ov::element::i32, ov::Shape({1}));
    auto const_val_2 = std::make_shared<ov::opset1::Constant>(ov::element::i32, ov::Shape({1}), std::vector<int>({2}));
    auto mulitply_then = std::make_shared<ov::opset1::Multiply>(inp_then, const_val_2);
    auto result_then = std::make_shared<ov::opset6::Result>(mulitply_then);
    auto model_then = std::make_shared<ov::Model>(ov::ResultVector({result_then}), ov::ParameterVector({inp_then}), "model_then");

    // If else
    auto inp_else = std::make_shared<ov::opset1::Parameter>(ov::element::i32, ov::Shape({1}));
    auto const_val_4 = std::make_shared<ov::opset1::Constant>(ov::element::i32, ov::Shape({1}), std::vector<int>({4}));
    auto mulitply_else = std::make_shared<ov::opset1::Multiply>(inp_else, const_val_4);
    auto result_else = std::make_shared<ov::opset6::Result>(mulitply_else);
    auto model_else = std::make_shared<ov::Model>(ov::ResultVector({result_else}), ov::ParameterVector({inp_else}), "model_else");

    if_op->set_then_body(model_then);
    if_op->set_else_body(model_else);

    if_op->set_input(input1->output(0), inp_then, inp_else);
    if_op->set_input(input2->output(0), inp_then, nullptr);
    if_op->set_input(input3->output(0), nullptr, inp_else);

    if_op->set_output(result_then, result_else);

    auto result = std::make_shared<ov::opset6::Result>(if_op);
    auto model = std::make_shared<ov::Model>(ov::ResultVector({result}), ov::ParameterVector({input1, input2, input3}), "model_if");

    return model;
}

bool test_model_if_device(std::string device)
{
    std::cout << " == Start test: " << __FUNCTION__ << " =======================" << std::endl;
    ov::Core core = ov::Core();

    // Initial input
    ov::Tensor input1 = initTensor<int>(ov::element::i32, ov::Shape({1}), std::vector<int>({0}));
    ov::Tensor input2 = initTensor<int>(ov::element::i32, ov::Shape({1}), std::vector<int>({2}));
    ov::Tensor input3 = initTensor<int>(ov::element::i32, ov::Shape({1}), std::vector<int>({4}));

    auto model = initIfModel();

    // Compile model.
    std::cout << " == compile_model" << std::endl;
    auto compiledModel = core.compile_model(model, device);
    std::cout << " == create_infer_request" << std::endl;
    auto inferRequest = compiledModel.create_infer_request();

    std::cout << " == set_input_tensors" << std::endl;
    inferRequest.set_input_tensors(0, {input1});
    inferRequest.set_input_tensors(1, {input2});
    inferRequest.set_input_tensors(2, {input3});

    // Inference
    std::cout << "  == Start infer. device=" << device << std::endl;
    inferRequest.infer();

    auto outTensor = inferRequest.get_output_tensor();

    // Check output
    const int *ouptputData = outTensor.data<int>();
    std::cout << "  == outShape = " << outTensor.get_shape() << ", ouptputData= " << ouptputData[0] << std::endl;
    std::cout << "Test model finish." << std::endl;
    return true;
}

bool test_model_if()
{
    test_model_if_device("CPU");
    // test_model_if_device("TEMPLATE");
    return true;
}
